### PR TITLE
Add Windows node reboot

### LINKED ISSUE #104428: Add tests for Windows Node Reboots
# Problem
As of now, when we reboot windows nodes, OVS goes down which breaks Antrea CNI.
I suspect reboots may have other effects on other services/CNIs in windows clusters.

Thus, we need a Reboot test case (azure has its own tests for node reboots, but theyre not upstreamed yet).

# Solution

So, lets add a `Distruptive` , `Slow` , `Serial` reboot test for `sig-windows` nodes. 

This is a good way for us to begin exersizing Priveliged containers.
However, if theres *another* way to get a windows node to reboot (from inside a container) we should explore it.

### PR DESCRIPTION
**Pull request by weiwenli97**
#### Which issue(s) this PR fixes:

Fixes #104428

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by weiwenli97** [comment]
Now I am testing the testcase. But I found Restart-computer failed with "Privilege not held". So I am trying to use shutdown.exe /r to test the reboot.

**Comment by jayunit100** [comment]
@weiwenli97 has some interesting research/updates ! feel free to paste here Wenli

**Comment by weiwenli97** [comment]
Deploy the Windows dev env in my local machine. I have to set VAGRANT_HTTP_PROXY, VAGRANT_HTTPS_PROXY and VAGRANT_NO_PROXY for the control plane setting up. Then I have to update controlplane.sh to add Environment paramters for I have to use proxy to access.

Also, my network is slow. I need use ctr.exe images pull some images in advance, such sandbox image and windows image.

**Comment by weiwenli97** [comment]
After the test env is set up, I have deployed two host-process pods, one is to test READ and the other one is test Write in the host. And both are successful and I check the host machine.

When I want to deploy the host-process pod to reboot the host, it failed for "Privilege not held" when I used Restart-Computer. Then I try "shutdown.exe /r /t 1" but it failed to reboot the host. And I am debugging on it.

**Comment by jayunit100** [comment]
@marosset -- do we expect HostProcess containers to be able to issue `reboot` or `restart-computer` or anything of the sort ? 

thats the most important premise for this test :)

File: test/e2e/windows/reboot_node.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("[Feature:Windows] RebootHost containers [Slow]", func() {
```
Comment (marosset) [inline test/e2e/windows/reboot_node.go:31]: Please add [Serial] and [Destructive] too

**Comment by marosset** [comment]
Lots of discussions  around this have been happening in slack:
https://kubernetes.slack.com/archives/C0SJ4AFB7/p1631668837117600

**Comment by spiffxp** [comment]
/test pull-kubernetes-node-e2e-containerd
ref: #105381

**Review by jsturtevant — CHANGES_REQUESTED** [review]
could we also schedule a hostprocess pod after the reboot and look up the reboot time of the machine to ensure the node was rebooted and the pod itself didn't just restart?

**Comment by marosset** [comment]
@weiwenli97 can you squash all the commits down to 1 commit for cleanliness (after other feedback gets addressed)

**Comment by ritazh** [comment]
/remove-sig auth
Feel free to add us if needed.

File: test/e2e/windows/reboot_node.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = SIGDescribe("[Feature:Windows] [Excluded:WindowsDocker] [MinimumKubeletVersion:1.22] RebootHost containers [Serial] [Disruptive] [Slow]", func() {
```
Comment (knabben) [inline test/e2e/windows/reboot_node.go:32]: not sure we need "whitespaces" between tags here. just to follow the standard.

File: test/e2e/windows/reboot_node.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+		framework.ExpectNoError(err, "Error retrieving pod")
+		framework.ExpectEqual(p.Status.Phase, v1.PodSucceeded)
+
+		agnPodOut, err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.TODO(), agnPod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "getting pod info before reboot")
+		preRebootPodIP := agnPodOut.Status.PodIP
+		preRebootPodStartTime := agnPodOut.Status.StartTime
+
+		ginkgo.By("Waiting 5 minutes")
```
[the remark this answers is not in the document]
Comment (jayunit100) [reply test/e2e/windows/reboot_node.go:106]: Given that this is working, could we merge it as a initial test (we'd selfishly like to use these downstream, without forking :))... and then improve the sleep time in *another* iteration on this MR ? @marosset

**Comment by jayunit100** [comment]
Talked with @ravisantoshgudimetla @jsturtevant  @claudiubelu : we want to use a pod as a sempahore to indicate node being up:

- schedule pod direct to a win node
- reboot that node
- continously curl the pod until the ep goes down
- once node back up, now a new family of tests can be written but this issue is done and we can do those in follow on issues

File: test/e2e/windows/reboot_node.go (hunk 2 above)
Comment (weiwenli97) [reply test/e2e/windows/reboot_node.go:34]: Yes. I have run ./hack/update_gofmt.sh and  it should follow the standard.

File: test/e2e/windows/reboot_node.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (weiwenli97) [reply test/e2e/windows/reboot_node.go]: Please help review the new codes. And I didn't sleep a static 5 minutes. I added a loop in it. Thanks!

**Comment by jsturtevant** [comment]
Thanks for making the changes.  They look like they would work. 

Sorry to be a pain but I was looking for something else and came across the command `systeminfo | find "System Boot Time` in the article https://infosharepoint.com/how-to-find-server-reboot-event/. Could we use that in a HPC to be a possibly be cleaner more reliable solution?

**Comment by dims** [comment]
@weiwenli97 can you please respond to the last comment from @jsturtevant ?

**Comment by weiwenli97** [comment]
@dims Thank you very much and I am very sorry to miss the comments.

@jsturtevant I am trying the solution. But I don't know what HPC refers here. Thanks!

**Comment by jsturtevant** [comment]
Sorry try to avoid the use of acronyms.  that should be host process container == hpc.

Was thinking we can query for the reboot time using a host process container.

**Comment by weiwenli97** [comment]
> 



> Sorry try to avoid the use of acronyms. that should be host process container == hpc.
> 
> Was thinking we can query for the reboot time using a host process container.

Yes. I could compare the "System Boot Time" before and after the host rebooted.
But it only tells us the host is rebooted and cannot tell us how many times it reboots.

**Comment by jayunit100** [comment]
`test/e2e/windows/reboot_node.go:37:3: undefined: SkipUnlessWindowsHostProcessContainersEnabled` fix that :) then re run 

then imo we should merge and iterate !

**Comment by aravindhp** [comment]
> > 
> 
> > Sorry try to avoid the use of acronyms. that should be host process container == hpc.
> > Was thinking we can query for the reboot time using a host process container.
> 
> Yes. I could compare the "System Boot Time" before and after the host rebooted. But it only tells us the host is rebooted and cannot tell us how many times it reboots.

FWIW, there is a hacky way to figure out number of times a system rebooted by doing `get-eventlog system | where-object {$_.eventid -eq 6005}`. This tells you the number of times the EventLog was restarted indirectly telling you the number of reboots. Ref: [How to get last 10 reboot times of a computer?](https://social.technet.microsoft.com/Forums/windowsserver/en-US/e47bc3a4-50d9-4ad8-aae0-2209098e4e71/how-to-get-last-10-reboot-times-of-a-computer?forum=winserverpowershell)

**Comment by weiwenli97** [comment]
> Thanks for making the changes. They look like they would work.
> 
> Sorry to be a pain but I was looking for something else and came across the command `systeminfo | find "System Boot Time` in the article https://infosharepoint.com/how-to-find-server-reboot-event/. Could we use that in a HPC to be a possibly be cleaner more reliable solution?

Yes. I have  added system boot time to check plus install time.
Could you please help review it for me? Thanks!

**Comment by aravindhp** [comment]
@weiwenli97 is it possible to squash merge all your comments? The intermediate merge commits does not add useful information.

**Comment by weiwenli97** [comment]
> 

I have checked last boot time and install time. Must we use event id to check?

**Comment by aravindhp** [comment]
Ignore my comment. I misread reboot time to mean number of reboots.

**Comment by weiwenli97** [comment]
> @weiwenli97 is it possible to squash merge all your comments? The intermediate merge commits does not add useful information.

When it is merged, it must be squashed.

**Comment by aravindhp** [comment]
> > @weiwenli97 is it possible to squash merge all your comments? The intermediate merge commits does not add useful information.
> 
> When it is merged, it must be squashed.

AFAIK, the tide will not squash merge

**Comment by jayunit100** [comment]
tide/merge-method-squash

**Comment by jayunit100** [comment]
Wenli can u rebase to 1 commit plz

**Comment by marosset** [comment]
/remove-sig apps architecture auth cli cloud-provider cluster-lifecycle network node release scheduling storage testing
/remove-wg structured-logging

(I'm guessing these all came in from a merge or something?)

**Comment by marosset** [comment]
/lgtm
@jsturtevant do you have any additional feedback? All of my concerns have been addressed.

**Comment by marosset** [comment]
/remove-area apiserver cloudprovider code-generation conformance dependency e2e-test-framework ipvs kubectl kubelet provider/gcp release-eng

**Pull request merged**